### PR TITLE
filter

### PR DESCRIPTION
**Pull request by MinhDo2007**

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -3,14 +3,14 @@ def new
   end
 
   def create
-    user = User.find_by email: params[:session][:email].downcase
+    user = User.find_by(email: params[:session][:email].downcase)
```
Comment (NguyenTanDuc) [inline app/controllers/sessions_controller.rb]: remove ()

Comment (NguyenTanDuc) [inline app/controllers/sessions_controller.rb]: remove ()

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -3,14 +3,14 @@ def new
...
     if user && user.authenticate(params[:session][:password])
       log_in user
       remember user
       params[:session][:remember_me] == "1" ? remember(user) : forget(user)
       redirect_to user
     else
-      flash[:error] = t "login.failed"
+      flash[:error] = t("login.failed")
```
Comment (NguyenTanDuc) [inline app/controllers/sessions_controller.rb]: remove (), and flash[:danger] not :error

File: app/models/word.rb [hunk 3]
```diff
@@ -2,4 +2,10 @@ class Word < ActiveRecord::Base
   has_many :results, dependent: :destroy
   has_many :answers, dependent: :destroy
   belongs_to :category
+
+  scope :learned, ->(user){where("id IN (SELECT word_id FROM results WHERE lesson_id
```
Comment (NguyenTanDuc) [inline app/models/word.rb]: remove () => user{where

File: app/controllers/sessions_controller.rb [hunk 4]
```diff
@@ -10,7 +10,7 @@ def create
       params[:session][:remember_me] == "1" ? remember(user) : forget(user)
       redirect_to user
     else
-      flash[:error] = t "login.failed"
+      flash[:danger] = t("login.failed")
```
Comment (nguyenngoctruong) [inline app/controllers/sessions_controller.rb]: remove ()

File: app/views/sessions/new.html.erb [hunk 5]
```diff
@@ -11,6 +11,11 @@
       <%= link_to t("forgot-link"), new_password_reset_path %>
       <%= f.password_field :password, class: "form-control" %>
 
+      <%= f.label :remember_me, class: "checkbox inline" do %>
+        <%= f.check_box :remember_me %>
+        <span>Remember me on this computer</span>
```
Comment (nguyenngoctruong) [inline app/views/sessions/new.html.erb]: I18n

File: app/views/shared/_search.html.erb [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    <p><%= t :categories %></p>
+    <div id="search-field">
+      <div class="col-md-5">
+        <%= collection_select(:filter, :category_id, @categories, :id,
+                              :name, {prompt: false},
+        {class: "form-control"}) %>
+      </div>
+    </div>
+    <div class="tm-radio">
+    <% ["all", "learned", "not_learned"].each do |item| %>
```
Comment (NguyenTanDuc) [inline app/views/shared/_search.html.erb]: indent

File: app/views/shared/_user_info.html.erb [hunk 7]
```diff
@@ -0,0 +1,2 @@
+<%= link_to gravatar_for(current_user, size: 50), current_user %>
```
Comment (phamkykhoi) [inline app/views/shared/_user_info.html.erb]: Move 50 to Settings.

File: app/controllers/words_controller.rb [hunk 8]
```diff
@@ -0,0 +1,19 @@
+class WordsController < ApplicationController
+  def index
+    @categories = Category.all
+    if params[:filter].present?
+      if params[:filter][:category_id].nil?
+        @category = Category.first
+      else
+        @category = Category.find params[:filter][:category_id]
```
Comment (phamkykhoi) [inline app/controllers/words_controller.rb]: Xử lý phức tạp

File: app/views/words/index.html.erb [hunk 9]
```diff
@@ -0,0 +1,15 @@
+<%= render "shared/search" %>
+<div class="container">
+  <table class="table table-bordered">
+    <tr>
+      <th>ID</th>
```
Comment (TrucLuu) [inline app/views/words/index.html.erb]: i18n

File: app/controllers/words_controller.rb [hunk 10]
```diff
@@ -0,0 +1,13 @@
+class WordsController < ApplicationController
+  def index
+    @categories = Category.all
+    if params[:filter].present? && params[:filter_state].present?
+      @category = Category.find params[:filter][:category_id]
+      @words = Word.filter_category @category
+      @words = @words.send params[:filter_state], current_user unless params[:filter_state]== "all"
```
Comment (nguyenngoctruong) [inline app/controllers/words_controller.rb]: add space before "=="

File: app/models/word.rb [hunk 11]
```diff
@@ -2,4 +2,10 @@ class Word < ActiveRecord::Base
   has_many :results, dependent: :destroy
   has_many :answers, dependent: :destroy
   belongs_to :category
+
+  scope :learned, ->user{where("id IN (SELECT word_id FROM results WHERE lesson_id 
+                                   IN (SELECT id FROM lessons WHERE user_id = ?))", user.id)}
```
Comment (vuongxuanhung) [inline app/models/word.rb]: Viết dư này cho gọn ;)
scope :learned, ->(user){where id: Result.select(word_id).where(lesson_id: user.lessons)}
query vẫn sẽ như vậy :)

File: app/models/word.rb [hunk 12]
```diff
@@ -2,4 +2,10 @@ class Word < ActiveRecord::Base
...
+  scope :not_learned, ->user{where("id NOT IN (SELECT word_id FROM results WHERE lesson_id 
+                                       IN (SELECT id FROM lessons WHERE user_id = ?))", user.id)}
```
Comment (vuongxuanhung) [inline app/models/word.rb]: scope :learned, ->(user){where.not id: Result.select(word_id).where(lesson_id: user.lessons)}

File: config/routes.rb [hunk 13]
```diff
@@ -15,4 +15,11 @@
   resources :account_activations, only: :edit
   resources :password_resets, except: [:destroy, :index]
   resources :categories, only: :index
+  resources :users do
+    member do
```
Comment (vuongxuanhung) [inline config/routes.rb]: vi phạm Restful :-"

File: app/controllers/words_controller.rb [hunk 14]
```diff
@@ -0,0 +1,13 @@
+class WordsController < ApplicationController
+  def index
+    @categories = Category.all
+    if params[:filter].present? && params[:filter_state].present?
+      @category = Category.find params[:filter][:category_id]
```
Comment (phamkykhoi) [inline app/controllers/words_controller.rb]: Error when hack link

File: app/views/words/index.html.erb [hunk 15]
```diff
@@ -0,0 +1,15 @@
+<%= render "shared/search" %>
+<div class="container">
+  <table class="table table-bordered">
+    <tr>
+      <th><%= t :ID %></th>
+      <th><%= t :word_content %></th>
+    </tr>
+    <% @words.each do |word| %>
```
Comment (phamkykhoi) [inline app/views/words/index.html.erb]: Use render @word

File: app/views/shared/_search.html.erb [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+    <p><%= t :categories %></p>
+    <div id="search-field">
+      <div class="col-md-5">
+        <%= select_tag "category_id", options_from_collection_for_select
+          (@categories, "id", "name", params[:category_id]), include_blank: t("word.all") %>
+      </div>
+    </div>
+    <div class="tm-radio">
+      <% ["all", "learned", "not_learned"].each do |item| %>
+        <%= label_tag item, class: 'radio-inline' do %>
```
Comment (khanhkhanh) [inline app/views/shared/_search.html.erb]: ' => "

File: app/controllers/words_controller.rb [hunk 17]
```diff
@@ -0,0 +1,14 @@
+class WordsController < ApplicationController
+  def index
+    @categories = Category.all
```
Comment (phamkykhoi) [inline app/controllers/words_controller.rb]: @categories = params[:category_id].present? ? Category.find params[:category_id] :  Category.all

File: app/models/word.rb [hunk 18]
```diff
@@ -6,4 +6,11 @@ class Word < ActiveRecord::Base
   accepts_nested_attributes_for :answers,
                                 reject_if: lambda {|a| a[:content].blank?},
                                 allow_destroy: true
+
+  scope :learned, ->user{where("id IN (SELECT word_id FROM results 
+  						  WHERE lesson_id IN (SELECT id FROM lessons WHERE user_id = ?))", user.id)}
+  scope :not_learned, ->user{where("id NOT IN (SELECT word_id FROM results 
+  							  WHERE lesson_id IN (SELECT id FROM lessons WHERE user_id = ?))", user.id)}
+  scope :filter_category, ->category{where(category: category) if category.present?}
```
Comment (dieunb) [inline app/models/word.rb]: remove `()`

File: app/views/shared/_search.html.erb [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+  <div class="row">
+    <p><%= t :categories %></p>
+    <div id="search-field">
+      <div class="col-md-5">
+       <%= select_tag "category_id", options_from_collection_for_select(@categories, "id", "name", params[:category_id]), 
+       include_blank: t("word.all") %>
+      </div>
+    </div>
+    <div class="tm-radio">
+      <% ["get_all", "learned", "not_learned"].each do |item| %>
```
Comment (dieunb) [inline app/views/shared/_search.html.erb]: Store in Setting `["get_all", "learned", "not_learned"].`

File: app/controllers/words_controller.rb [hunk 20]
```diff
@@ -0,0 +1,10 @@
...
+    @words = if params[:filter_state].present?
+      Word.filter_category(params[:category_id]).send params[:filter_state], current_user
+    else
+      Word.all
```
Comment (phamkykhoi) [inline app/controllers/words_controller.rb]: Indent

**Pull request merged**